packages/jaeger-ui/src/reducers/deep-dependency-graph.tsx

```diff
@@ -0,0 +1,141 @@
+  });
+  return {
+    ...state,
+    [key]: {
+      ...stateEntry,
+      styleStates,
+    },
+  };
+}
+
+export function clearStyleState(state: TDdgState, { payload }: { payload: TDdgClearStylePayload }) {
+  const { service, operation, start, end, visibilityIndices, style } = payload;
+  const key = stateKey(service, operation, start, end);
+  const stateEntry: TDdgStateEntry | undefined = state[key];
+  if (!stateEntry || stateEntry.state !== fetchedState.DONE) {
+    console.warn('Cannot change style state for unloaded Deep Dependency Graph'); // eslint-disable-line no-console
+    return state;
+  }
+  const styleStates = new Map(stateEntry.styleStates);
+  (visibilityIndices || Array.from(styleStates.keys())).forEach(idx => {
+    if (style == null) {
+      styleStates.delete(idx);
+    } else {
+      styleStates.set(idx, (styleStates.get(idx) || 0) & ~style); // eslint-disable-line no-bitwise
+      if (styleStates.get(idx) === 0) styleStates.delete(idx);
+    }
+  });
+  return {
+    ...state,
+    [key]: {
+      ...stateEntry,
+      styleStates,
+    },
+  };
+}
+
+export function fetchDeepDependencyGraphStarted(state: TDdgState, { meta }: { meta: TDdgActionMeta }) {
+  const { query } = meta;
+  const { service, operation, start, end } = query;
+  const key = stateKey(service, operation, start, end);
+  return {
+    ...state,
+    [key]: {
+      state: fetchedState.LOADING,
+    },
+  };
+}
+
+export function fetchDeepDependencyGraphDone(
+  state: TDdgState,
+  { meta, payload }: { meta: TDdgActionMeta; payload: TDdgPayload }
+) {
+  const { query } = meta;
+  const { service, operation, start, end } = query;
+  const key = stateKey(service, operation, start, end);
+  return {
+    ...state,
+    [key]: {
+      model: transformDdgData(payload, { service, operation }),
+      state: fetchedState.DONE,
+      styleStates: new Map(),
+    },
+  };
+}
+
+export function fetchDeepDependencyGraphErred(
+  state: TDdgState,
+  { meta, payload }: { meta: TDdgActionMeta; payload: ApiError }
+) {
+  const { query } = meta;
+  const { service, operation, start, end } = query;
+  const key = stateKey(service, operation, start, end);
+  return {
+    ...state,
+    [key]: {
+      error: payload,
+      state: fetchedState.ERROR,
+    },
+  };
+}
+
+export default handleActions(
+  {
+    [`${fetchDeepDependencyGraph}_PENDING`]: fetchDeepDependencyGraphStarted,
+    [`${fetchDeepDependencyGraph}_FULFILLED`]: guardReducerWithMeta<TDdgState, TDdgPayload, TDdgActionMeta>(
+      fetchDeepDependencyGraphDone
+    ),
+    [`${fetchDeepDependencyGraph}_REJECTED`]: guardReducerWithMeta<TDdgState, ApiError, TDdgActionMeta>(
+      fetchDeepDependencyGraphErred
+    ),
+
+    [actionTypes.ADD_STYLE_STATE]: guardReducer<TDdgState, { payload: TDdgAddStylePayload }>(addStyleState),
+    [actionTypes.CLEAR_STYLE_STATE]: guardReducer<TDdgState, { payload: TDdgClearStylePayload }>(
+      clearStyleState
+    ),
+  },
```

Comment:
I'm surprised to see the type paramters are necessary. Is it the result of `handleActions`? The following seems to work:

[playground](https://www.typescriptlang.org/play/index.html#src=type%20Action%3CT%3E%20%3D%20%7B%20payload%3A%20T%20%7D%3B%0D%0A%0D%0Atype%20TDdgState%20%3D%20%7B%7D%3B%0D%0Atype%20TDdgActionMeta%20%3D%20%7B%7D%3B%0D%0Atype%20ApiError%20%3D%20%7B%7D%3B%0D%0A%0D%0Afunction%20guardReducerWithMeta%3CTState%2C%20TPayload%2C%20TMeta%3E(%0D%0A%20%20fn%3A%20(state%3A%20TState%2C%20action%3A%20%7B%20meta%3A%20TMeta%3B%20payload%3A%20TPayload%20%7D)%20%3D%3E%20TState%0D%0A)%20%7B%0D%0A%20%20return%20function%20reducer(state%3A%20TState%2C%20action%3A%20Action%3CTPayload%3E%20%26%20%7B%20meta%3A%20TMeta%20%7D)%20%7B%0D%0A%20%20%20%20if%20(!action.payload%20%7C%7C%20!action.meta)%20%7B%0D%0A%20%20%20%20%20%20return%20state%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20return%20fn(state%2C%20action%20as%20%7B%20meta%3A%20TMeta%3B%20payload%3A%20TPayload%20%7D)%3B%0D%0A%20%20%7D%3B%0D%0A%7D%0D%0A%0D%0Afunction%20fetchDeepDependencyGraphErred(%0D%0A%20%20state%3A%20TDdgState%2C%0D%0A%20%20%7B%20meta%2C%20payload%20%7D%3A%20%7B%20meta%3A%20TDdgActionMeta%3B%20payload%3A%20ApiError%20%7D%0D%0A)%20%7B%0D%0A%20%20%20%20console.log(state%2C%20meta%2C%20payload)%3B%0D%0A%20%20%20%20return%20state%3B%0D%0A%7D%0D%0A%0D%0Aconsole.log(guardReducerWithMeta(fetchDeepDependencyGraphErred))%3B%0D%0A)

```ts
type Action<T> = { payload: T };

type TDdgState = {};
type TDdgActionMeta = {};
type ApiError = {};

function guardReducerWithMeta<TState, TPayload, TMeta>(
  fn: (state: TState, action: { meta: TMeta; payload: TPayload }) => TState
) {
  return function reducer(state: TState, action: Action<TPayload> & { meta: TMeta }) {
    if (!action.payload || !action.meta) {
      return state;
    }
    return fn(state, action as { meta: TMeta; payload: TPayload });
  };
}

function fetchDeepDependencyGraphErred(
  state: TDdgState,
  { meta, payload }: { meta: TDdgActionMeta; payload: ApiError }
) {
    console.log(state, meta, payload);
    return state;
}

console.log(guardReducerWithMeta(fetchDeepDependencyGraphErred));

```

Reply:
I believe it was necessitated by `handleActions`
I don't think it should be done in these PRs, but I think it would probably be best to migrate away from the `redux-actions` package as it is just a bit of sugar that comes with some headaches and typing issues.